### PR TITLE
Discover and debug Flutter apps that were not launched from Atom

### PR DESCRIPTION
**Pull request by jason-simmons**

### CODE REVIEW

**Comment by jason-simmons** [comment]
@devoncarew

File: lib/flutter/flutter_connect.dart [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+final Logger _logger = new Logger('flutter.connect');
+
+/// Connect the tools to a Flutter app that is already running on a device.
+class FlutterConnectManager implements Disposable {
+  Disposables _disposables = new Disposables();
+
+  ConnectDialog connectDialog;
+
+  FlutterConnectManager() {
```
Comment (devoncarew) [inline lib/flutter/flutter_connect.dart]: nit: you can omit this ctor, it's implied

File: lib/flutter/flutter_connect.dart [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  void dispose() => _disposables.dispose();
+}
+
+class ConnectDialog implements Disposable {
+  TitledModelDialog dialog;
+  CoreElement _listGroup;
+
+  ConnectDialog() {
+    dialog = new TitledModelDialog('Connect to Flutter App', classes: 'jobs-dialog');
```
Comment (devoncarew) [inline lib/flutter/flutter_connect.dart]: We probably want to change the css name (and its uses) from `jobs-dialog` to something more general like `list-dialog`. Can you add a TODO? (I'll follow it up)

File: lib/flutter/flutter_daemon.dart [hunk 3]
```diff
@@ -340,6 +340,13 @@ class DaemonDomain extends Domain {
   Future shutdown() => _call('daemon.shutdown');
 }
 
+/// Describes an app running on the device.
+class DiscoveredApp {
+  DiscoveredApp(this.id, this.observatoryPort);
+  String id;
```
Comment (devoncarew) [inline lib/flutter/flutter_daemon.dart]: `final`

File: lib/flutter/flutter_daemon.dart [hunk 4]
```diff
@@ -393,6 +408,22 @@ class DeviceDomain extends Domain {
   Future enable() => _call('device.enable');
 
   Future disable() => _call('device.disable');
+
+  Future<int> forward(String deviceId, int devicePort, int hostPort) {
```
Comment (devoncarew) [inline lib/flutter/flutter_daemon.dart]: I think we can make the hostPort optional here (`, [int hostPort]`).

File: lib/flutter/flutter_daemon.dart [hunk 5]
```diff
@@ -393,6 +408,22 @@ class DeviceDomain extends Domain {
...
+    return _call('device.forward', _stripNullValues({
+      'deviceId': deviceId,
+      'devicePort': devicePort,
+      'hostPort': hostPort,
+    })).then((Map<String, dynamic> result) => result['hostPort']);
+  }
+
+  Future unforward(String deviceId, int devicePort, int hostPort) {
```
Comment (devoncarew) [inline lib/flutter/flutter_daemon.dart]: Is hostPort required for unforwarding?

File: lib/flutter/flutter_launch.dart [hunk 6]
```diff
@@ -210,11 +233,48 @@ class _LaunchInstance {
...
+      killHandler: _kill,
+      cwd: project.path,
+      title: description,
+      targetName: _device?.name
+    );
+    launchManager.addLaunch(_launch);
+  }
+
+  Future<Launch> launch() async {
+    _observatoryPort = await _daemon.device.forward(_device.id, _observatoryDevicePort, null);
```
Comment (devoncarew) [inline lib/flutter/flutter_launch.dart]: We can omit `null` here once the last param is optional.

File: lib/flutter/flutter_tools.dart [hunk 7]
```diff
@@ -40,6 +43,14 @@ class FlutterToolsManager implements Disposable {
         'flutter:upgrade',
         _upgrade
       ));
+      disposables.add(atom.commands.add(
+        'atom-workspace',
+        'flutter:connect',
```
Comment (devoncarew) [inline lib/flutter/flutter_tools.dart]: The command name will show up in the command palette dialog as `Flutter: Connect`. We should give it a name that indicates a bit more what it's for - something like `flutter:connect-remote-debugger`.

File: lib/flutter/flutter_connect.dart [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+      atom.workspace.getActiveTextEditor()?.getPath());
+    if (project == null) {
+      atom.notifications.addWarning('No active project.');
+      return;
+    }
+
+    List<LaunchConfiguration> configs = project == null ?
+        [] : launchConfigurationManager.getConfigsForProject(project.path);
+
+    if (configs.isNotEmpty) {
```
Comment (devoncarew) [inline lib/flutter/flutter_connect.dart]: It looks like the connection will not proceed if there's not an existing launch config; we should create one in this case. Can you open an issue for this? I'm happy in implement it.

**Comment by devoncarew** [comment]
Fantastic! Some comments, and probably some follow-up work (w/ css names, and creating an initial launch config if there isn't already an existing one). lgtm

File: lib/flutter/flutter_connect.dart (hunk 1 above)
Comment (jason-simmons) [inline lib/flutter/flutter_connect.dart]: done

File: lib/flutter/flutter_connect.dart (hunk 2 above)
Comment (jason-simmons) [inline lib/flutter/flutter_connect.dart]: done

File: lib/flutter/flutter_daemon.dart (hunk 3 above)
Comment (jason-simmons) [inline lib/flutter/flutter_daemon.dart]: done

File: lib/flutter/flutter_daemon.dart (hunk 4 above)
Comment (jason-simmons) [inline lib/flutter/flutter_daemon.dart]: done

File: lib/flutter/flutter_daemon.dart (hunk 5 above)
Comment (jason-simmons) [inline lib/flutter/flutter_daemon.dart]: The DevicePortForwarder.unforward interface in flutter_tools takes a ForwardedPort that includes both device and host ports.  So an implementation might need the host port.

File: lib/flutter/flutter_launch.dart (hunk 6 above)
Comment (jason-simmons) [inline lib/flutter/flutter_launch.dart]: done

File: lib/flutter/flutter_tools.dart (hunk 7 above)
Comment (jason-simmons) [inline lib/flutter/flutter_tools.dart]: done

File: lib/flutter/flutter_connect.dart (hunk 8 above)
Comment (jason-simmons) [inline lib/flutter/flutter_connect.dart]: done - https://github.com/dart-atom/dartlang/issues/1023

**Comment by devoncarew** [comment]
lgtm! Can you merge this to master - the compiled JS files have since been updated in master. I can then merge this.

**Comment by devoncarew** [comment]
Merging - thanks!

**Pull request merged**